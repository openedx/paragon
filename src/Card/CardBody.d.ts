import React from 'react';

export interface CardBodyProps {
  className?: string,
  children?: React.ReactNode,
}
declare const CardBody: React.ForwardRefExoticComponent<CardBodyProps>;
export default CardBody;
