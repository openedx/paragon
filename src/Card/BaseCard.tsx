import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import type { ComponentWithAsProp, BsPropsWithAs } from '../utils/types/bootstrap';

// @ts-ignore
import CardBody from './CardBody';

const BASE_CARD_CLASSNAME = 'card';

const colorVariants = [
  'primary',
  'secondary',
  'success',
  'danger',
  'warning',
  'info',
  'dark',
  'light',
] as const;

const textVariants = [
  'white',
  'muted',
] as const;

type ColorVariant = typeof colorVariants[number];
type TextVariant = typeof textVariants[number];
interface Props extends BsPropsWithAs {
  prefix?: string;
  bgColor?: ColorVariant;
  textColor?: ColorVariant | TextVariant;
  borderColor?: ColorVariant;
  hasBody?: boolean;
  className?: string;
  children: React.ReactNode;
}
type BaseCardType = ComponentWithAsProp<'div', Props>;

const BaseCard : BaseCardType = React.forwardRef<HTMLDivElement, Props>(
  (
    {
      prefix,
      className,
      bgColor,
      textColor,
      borderColor,
      hasBody = false,
      children,
      as: Component = 'div',
      ...props
    },
    ref,
  ) => {
    const classes = classNames(
      className,
      prefix ? `${prefix}-${BASE_CARD_CLASSNAME}` : BASE_CARD_CLASSNAME,
      bgColor && `bg-${bgColor}`,
      textColor && `text-${textColor}`,
      borderColor && `border-${borderColor}`,
    );

    return (
      <Component ref={ref} {...props} className={classes}>
        {hasBody ? <CardBody>{children}</CardBody> : children}
      </Component>
    );
  },
);

/* eslint-disable react/require-default-props */
BaseCard.propTypes = {
  /** Prefix for component CSS classes. */
  prefix: PropTypes.string,
  /** Background color of the card. */
  bgColor: PropTypes.oneOf(colorVariants),
  /** Text color of the card. */
  textColor: PropTypes.oneOf([...colorVariants, ...textVariants]),
  /** Border color of the card. */
  borderColor: PropTypes.oneOf(colorVariants),
  /** Determines whether the card should render its children inside a `CardBody` wrapper. */
  hasBody: PropTypes.bool,
  /** Set a custom element for this component. */
  as: PropTypes.elementType,
  /** Additional CSS class names to apply to the card element. */
  className: PropTypes.string,
  /** The content to render inside the card. */
  children: PropTypes.node,
};

export default BaseCard;
