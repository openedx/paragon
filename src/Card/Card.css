/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:19 GMT
 */

:root {
  --pgn-spacing-card-focus-border-offset: 5px;
  --pgn-spacing-card-logo-bottom-offset-horizontal: .4375rem;
  --pgn-spacing-card-logo-bottom-offset-base: 1rem;
  --pgn-spacing-card-logo-left-offset-horizontal: .4375rem;
  --pgn-spacing-card-logo-left-offset-base: 1.5rem;
  --pgn-spacing-card-loading-skeleton-spacer: .313rem;
  --pgn-spacing-card-footer-action-gap: .5rem;
  --pgn-spacing-card-columns-gap: 1.25rem;
  --pgn-spacing-card-columns-count: 3;
  --pgn-spacing-card-margin-group: 12px;
  --pgn-spacing-card-spacer-y: .75rem;
  --pgn-spacing-card-spacer-x: 1.25rem;
  --pgn-size-card-logo-height: 4.125rem;
  --pgn-size-card-logo-width: 7.25rem;
  --pgn-size-card-image-vertical-max-height: 140px;
  --pgn-size-card-image-horizontal-width-max: 240px;
  --pgn-size-card-focus-border-radius: 5px;
  --pgn-size-card-focus-border-width: 2px;
  --pgn-size-card-border-radius-logo: .25rem;
  --pgn-size-card-border-radius-image: .3125rem;
  --pgn-typography-footer-text-font-size: var(--pgn-typography-font-size-small-x);
  --pgn-spacing-card-columns-margin: var(--pgn-spacing-card-spacer-y);
  --pgn-spacing-card-margin-grid: var(--pgn-spacing-card-margin-group);
  --pgn-spacing-card-margin-deck: var(--pgn-spacing-card-margin-group);
  --pgn-size-card-image-horizontal-width-min: var(--pgn-size-card-image-horizontal-width-max);
  --pgn-size-card-border-radius-base: var(--pgn-size-border-radius-base);
  --pgn-size-card-border-width: var(--pgn-size-border-width);
  --pgn-spacing-card-margin-grid-bottom: var(--pgn-spacing-spacer-3);
  --pgn-spacing-card-margin-deck-bottom: var(--pgn-spacing-spacer-3);
}
