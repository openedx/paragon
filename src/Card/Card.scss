@import "variables";
@import "~bootstrap/scss/card";

.pgn__card-grid {
  .row > div[class*="col-"] {
    margin-bottom: $card-grid-margin;
    display: flex;
    flex: 1 0 auto;
  }
}

.pgn__card-header {
  padding: 0 $card-spacer-x;
  display: flex;
  justify-content: space-between;

  .pgn__card-header-content {
    display: flex;
    flex-direction: column;
    margin-top: map_get($spacers, 4);
    overflow: auto;
  } 

  .pgn__card-header-title {
    color: $black;
    font-weight: $font-weight-bold;
  }
  .pgn__card-header-title-sm {
    @extend .pgn__card-header-title;
    font-size: $h4-font-size;
  }
  .pgn__card-header-title-md {
    @extend .pgn__card-header-title;
    font-size: $h3-font-size;
  }

  .pgn__card-header-subtitle {
    color: $gray-700;
    margin-top: map_get($spacers, 1);
  }
  .pgn__card-header-subtitle-sm {
    @extend .pgn__card-header-subtitle;
    font-size: $h5-font-size;
  }
  .pgn__card-header-subtitle-md {
    @extend .pgn__card-header-subtitle;
    font-size: $h4-font-size;
  }

  .pgn__card-header-actions {
    margin-top: $spacer;
    margin-left: $spacer;
    flex-shrink: 0;
  }
}

.pgn__card-divider {
  border-top: 1px solid $card-divider-bg;
  height: 0;
}

.pgn__card-section {
  padding: $card-spacer-x;
  background-color: $white;

  &.is-muted {
    background-color: $light-200;
  }

  .pgn__card-section-title {
    color: $black;
    font-weight: $font-weight-bold;
    font-size: $h5-font-size;
    margin-bottom: $card-spacer-y;
  }

  .pgn__card-section-actions {
    display: flex;
    justify-content: flex-end;
    margin-top: $card-spacer-y;
  }
}

.pgn__card-footer {
  @extend .w-100;
  padding: $card-spacer-x;
  background-color: $white;
  display: flex;
  align-items: center;
  justify-content: center;

  &.vertical {
    justify-content: flex-end;
    padding-top: 0;
    & > * + * {
      margin-left: $card-footer-actions-gap;
    }
  }

  &.vertical-stacked {
    flex-direction: column-reverse;

    & > * {
      width: 100%;
    }

    & > * + * {
      margin-top: $card-footer-actions-gap;
    }
  }

  &.horizontal {
    flex-wrap: wrap;

    & > :not(:last-child) {
      margin-right: $card-footer-actions-gap;
    }
  }

  &.horizontal-stacked {
    flex-direction: column-reverse;

    & > * + * {
      margin-top: $card-footer-actions-gap;
    }
  }
}

.pgn__card-footer-text {
  color: $gray-700;
  order: 1;

  &.vertical {
    margin-right: auto;
    order: -1;
  }

  &.vertical-stacked {
    margin-right: auto;
  }

  &.horizontal {
    margin-top: $card-spacer-x;
    width: 100%;
    text-align: center;
  }

  &.horizontal-stacked {
    margin-top: $card-spacer-x;
    width: 100%;
    text-align: center;
    order: -1;
  }
}

.pgn__card-wrapper-image-cap {
  position: relative;

  &.horizontal {
    max-width: 30%;

    .pgn__card-image-cap {
      height: 100%;
      border-top-left-radius: $card-border-radius;
      border-bottom-left-radius: $card-border-radius;
      border-top-right-radius: 0;
      max-width: -moz-available;          /* WebKit-based browsers will ignore this. */
      max-width: -webkit-fill-available;  /* Mozilla-based browsers will ignore this. */
      max-width: stretch;
      width: auto;
    }

    .pgn__card-logo-cap {
      left: $card-logo-left-offset-horizontal;
      bottom: $card-logo-bottom-offset-horizontal;
    }
  }
}

.pgn__card-image-cap {
  width: 100%;
  border-top-left-radius: $card-border-radius;
  border-top-right-radius: $card-border-radius;
}

.pgn__card-logo-cap {
  position: absolute;
  left: $card-logo-left-offset;
  bottom: #{-$card-logo-bottom-offset};
  width: $card-logo-width;
  height: $card-logo-height;
  border-radius: $card-logo-border-radius;
  box-shadow: $level-1-box-shadow;
  z-index: 1;
}

.card {
  &:focus,
  &.focus {
    box-shadow: $level-1-box-shadow;
    outline: none;

    &::before {
      content: '';
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      border: $card-border-width solid $card-border-focus-color;
      border-radius: $card-border-radius;
    }
  }

  &:hover {
    box-shadow: $level-1-box-shadow;
  }
}

// horizontal styles

.pgn__card.horizontal {
  flex-direction: row;
}
