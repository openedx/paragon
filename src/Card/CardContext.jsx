import React, { createContext } from 'react';
import PropTypes from 'prop-types';

const CardContext = createContext({});

const CardContextProvider = ({
  horizontal, children,
}) => (
  <CardContext.Provider value={{ horizontal }}>
    {children}
  </CardContext.Provider>
);

CardContextProvider.propTypes = {
  /** Display `horizontal` layout of the `Card` when the prop is set to `true`. */
  horizontal: PropTypes.bool,
  /** Specifies content of the component. */
  children: PropTypes.node,
};

CardContextProvider.defaultProps = {
  horizontal: false,
  children: null,
};

export { CardContextProvider };
export default CardContext;
