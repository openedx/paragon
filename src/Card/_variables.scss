// Card

$card-spacer-y:                      .75rem !default;
$card-spacer-x:                      1.25rem !default;
$card-border-width:                  $border-width !default;
$card-border-radius:                 $border-radius !default;
$card-border-color:                  rgba($black, .125) !default;
$card-border-focus-color:            rgba($black, .5) !default;
$card-border-focus-color-dark:       theme-color("primary", "focus") !default;
$card-inner-border-radius:           subtract($card-border-radius, $card-border-width) !default;
$card-cap-bg:                        rgba($black, .03) !default;
$card-cap-color:                     null !default;
$card-height:                        null !default;
$card-color:                         null !default;
$card-bg:                            $white !default;
$card-bg-dark:                       $primary-500 !default;
$card-bg-muted:                      $light-200 !default;

$card-img-overlay-padding:           1.25rem !default;

$card-group-margin:                  calc($grid-gutter-width / 2) !default;
$card-deck-margin:                   $card-group-margin !default;
$card-grid-margin:                   $card-group-margin !default;
$card-deck-margin-bottom:            map_get($spacers, 3) !default;
$card-grid-margin-bottom:            map_get($spacers, 3) !default;

$card-columns-count:                 3 !default;
$card-columns-gap:                   1.25rem !default;
$card-columns-margin:                $card-spacer-y !default;

$card-divider-bg:                    $light-400 !default;
$card-divider-margin-y:              $card-spacer-y !default;

$card-footer-actions-gap:            .5rem !default;

$card-logo-left-offset:              1.5rem !default;
$card-logo-bottom-offset:            1rem !default;
$card-logo-left-offset-horizontal:   .4375rem !default;
$card-logo-bottom-offset-horizontal: .4375rem !default;

$card-logo-width:                    7.25rem !default;
$card-logo-height:                   4.125rem !default;

$card-image-border-radius:           .3125rem !default;
$card-logo-border-radius:            .25rem !default;

$card-footer-text-font-size:         $x-small-font-size;

$card-image-horizontal-max-width:    240px !default;
$card-image-horizontal-min-width:    $card-image-horizontal-max-width !default;
$card-image-vertical-max-height:     140px !default;
$loading-skeleton-spacer:            .313rem !default;

$card-focus-border-offset:           5px !default;
$card-focus-border-width:            2px !default;
$card-focus-border-radius:           5px !default;
