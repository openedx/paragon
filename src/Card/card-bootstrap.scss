.card {
  display: flex;
  flex-direction: column;
  min-width: 0;
  word-wrap: break-word;
  background-color: $card-bg;
  background-clip: border-box;
  border: $card-border-width solid $card-border-color;
  position: relative;

  @include border-radius($card-border-radius);

  > hr {
    margin-right: 0;
    margin-left: 0;
  }

  > .list-group {
    border-top: inherit;
    border-bottom: inherit;

    &:first-child {
      border-top-width: 0;

      @include border-top-radius($card-inner-border-radius);
    }

    &:last-child {
      border-bottom-width: 0;

      @include border-bottom-radius($card-inner-border-radius);
    }
  }

  > .card-header + .list-group,
  > .list-group + .card-footer {
    border-top: 0;
  }
}

.card-body {
  flex: 1 1 auto;
  min-height: 1px;
  padding: $card-spacer-x;
  color: $card-color;
}

// Card deck
.card-deck {
  .card {
    margin-bottom: $card-deck-margin;
  }

  @include media-breakpoint-up(sm) {
    display: flex;
    flex-flow: row wrap;
    margin-right: calc(#{$card-deck-margin} * -1);
    margin-left: calc(#{$card-deck-margin} * -1);

    .card {
      flex: 1 0 0%;
      margin-right: $card-deck-margin;
      margin-bottom: 0; // Override the default
      margin-left: $card-deck-margin;
    }
  }
}

// Card groups
.card-group {
  > .card {
    margin-bottom: $card-group-margin;
  }

  @include media-breakpoint-up(sm) {
    display: flex;
    flex-flow: row wrap;

    > .card {
      flex: 1 0 0%;
      margin-bottom: 0;

      + .card {
        margin-left: 0;
        border-left: 0;
      }

      // Handle rounded corners
      @if $enable-rounded {
        &:not(:last-child) {
          @include border-right-radius(0);

          .card-img-top,
          .card-header {
            border-top-right-radius: 0;
          }

          .card-img-bottom,
          .card-footer {
            border-bottom-right-radius: 0;
          }
        }

        &:not(:first-child) {
          @include border-left-radius(0);

          .card-img-top,
          .card-header {
            border-top-left-radius: 0;
          }

          .card-img-bottom,
          .card-footer {
            border-bottom-left-radius: 0;
          }
        }
      }
    }
  }
}

// Columns
.card-columns {
  .card {
    margin-bottom: $card-columns-margin;
  }

  @include media-breakpoint-up(sm) {
    column-count: $card-columns-count;
    column-gap: $card-columns-gap;
    orphans: 1;
    widows: 1;

    .card {
      display: inline-block;
      width: 100%;
    }
  }
}

// Accordion
.accordion {
  overflow-anchor: none;

  > .card {
    overflow: hidden;

    &:not(:last-of-type) {
      border-bottom: 0;

      @include border-bottom-radius(0);
    }

    &:not(:first-of-type) {
      @include border-top-radius(0);
    }

    > .card-header {
      @include border-radius(0);

      margin-bottom: calc(#{$card-border-width} * -1);
    }
  }
}
