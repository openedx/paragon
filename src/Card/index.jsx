import React from 'react';
import BaseCard from 'react-bootstrap/Card';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import CardContext, { CardContextProvider } from './CardContext';
import CardHeader from './CardHeader';
import CardDivider from './CardDivider';
import CardSection from './CardSection';
import CardFooter from './CardFooter';
import CardImageCap from './CardImageCap';
import CardBody from './CardBody';

const Card = ({ horizontal, ...props }) => (
  <CardContextProvider horizontal={horizontal}>
    <BaseCard className={classNames('pgn__card', { horizontal })} {...props} />
  </CardContextProvider>
);

export { default as CardColumns } from 'react-bootstrap/CardColumns';
export { default as CardDeck } from 'react-bootstrap/CardDeck';
export { default as CardImg } from 'react-bootstrap/CardImg';
export { default as CardGroup } from 'react-bootstrap/CardGroup';
export { default as CardGrid } from './CardGrid';

Card.propTypes = {
  ...BaseCard.propTypes,
  /** Display `horizontal` layout of the `Card` when the prop is set to `true`. */
  horizontal: PropTypes.bool,
};

Card.defaultProps = {
  ...BaseCard.defaultProps,
  horizontal: false,
};

CardContextProvider.defaultProps = {
  horizontal: false,
};

Card.Header = CardHeader;
Card.Divider = CardDivider;
Card.Section = CardSection;
Card.Footer = CardFooter;
Card.ImageCap = CardImageCap;
Card.Context = CardContext;
Card.Body = CardBody;
export default Card;
