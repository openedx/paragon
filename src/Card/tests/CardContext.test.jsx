import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import { CardContextProvider } from '../CardContext';
import Card from '../index';

describe('correct rendering', () => {
  it('renders a context element', () => {
    const tree = renderer.create((
      <CardContextProvider>Text</CardContextProvider>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a context with correct props', () => {
    const wrapper = shallow((
      <Card horizontal />
    ));
    const contextProvider = wrapper.find(CardContextProvider);
    expect(contextProvider.props().horizontal).toBe(true);
  });
});
