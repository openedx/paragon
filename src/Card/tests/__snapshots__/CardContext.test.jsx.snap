// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correct rendering renders a context element 1`] = `"Text"`;
